### PR TITLE
Basic screens implementation

### PR DESCRIPTION
**Pull request by loucass003**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -5,7 +5,7 @@
   "main": "src/index.js",
   "type": "module",
   "scripts": {
-    "start": "node src/index.js | pino-human",
+    "start": "node --inspect=0.0.0.0 src/index.js | pino-human",
```
Comment (DeltaEvo) [inline package.json:8]: Maybe add a `debug` script for that or remove it :)

File: src/index.js [hunk 2]
```diff
@@ -44,11 +45,16 @@ const initial_world = [
...
+  register_screen({
+    id: 'player_screen',
+    size: { width: 8, height: 4 },
+  }),
 ].reduce((world, fn) => fn(world), {
   ...floor1,
   events: new EventEmitter(),
   next_entity_id: 0,
   next_window_id: 1, // 0 is the player inventory
+  screens: [],
```
Comment (DeltaEvo) [inline src/index.js:57]: This should not go here, can you not create it inside `register_screen` ?

File: src/index.js [hunk 3]
```diff
@@ -62,7 +68,7 @@ const initial_state = ({ entity_id, world }) => ({
     37: { type: 'bronze_coin', count: 10 },
     38: { type: 'menitrass_100', count: 1 },
   }),
-  game_mode: 2,
+  game_mode: 1,
```
Comment (DeltaEvo) [inline src/index.js:71]: Revert this change :)

File: src/login.js [hunk 4]
```diff
@@ -1,12 +1,22 @@
...
+  get_screen,
+  spawn_screen,
+  update_screen,
+} from './screen.js'
+
+const { createCanvas } = canvas
 
 export default function login({ client, events }) {
-  events.once('state', (state) => {
+  events.once('state', async (state) => {
```
Comment (DeltaEvo) [inline src/login.js:19]: async dosen't seems to be used

File: src/login.js [hunk 5]
```diff
@@ -71,7 +81,42 @@ export default function login({ client, events }) {
       title: { text: 'AresRPG' },
       fadeIn: 5,
       fadeOut: 2,
-      stay: 10,
+      stay: 1,
     })
+
+    const screen_pos = { ...world.spawn_position }
+    screen_pos.y += 15
```
Comment (DeltaEvo) [inline src/login.js:88]:
```suggestion
    const screen_pos = { ...world.spawn_position, y: world.spawn_position.y + 15 }
```

File: src/login.js [hunk 6]
```diff
@@ -71,7 +81,42 @@ export default function login({ client, events }) {
...
+        direction: { x: 1, y: 0, z: 0 },
+      }
+    )
+
+    let last_frame = null
+    const { size } = get_screen(world, 'player_screen')
+    const canvas = createCanvas(size.width * 128, size.height * 128)
+    const ctx = canvas.getContext('2d')
+    ctx.fillStyle = 'black'
+    ctx.fillRect(0, 0, size.width * 128, size.height * 128)
```
Comment (DeltaEvo) [inline src/login.js:104]: Maybe we should introduce a create_screen_canvas for that ?

File: src/login.js [hunk 7]
```diff
@@ -71,7 +81,42 @@ export default function login({ client, events }) {
...
+        ctx.ellipse(x, y, 10, 10, 0, 0, Math.PI * 2)
+        ctx.fill()
+        update_screen(
+          { client, world },
+          { screen_id, newCanvas: canvas, oldCanvas: last_frame }
+        )
+        last_frame = copy_canvas(canvas)
+      },
+      100
+    )
```
Comment (DeltaEvo) [inline src/login.js:120]:
```suggestion
    const { size } = get_screen(world, 'player_screen')
    const canvas = createCanvas(size.width * 128, size.height * 128)
   
    const ctx = canvas.getContext('2d')
    ctx.fillStyle = 'black'
    ctx.fillRect(0, 0, size.width * 128, size.height * 128)
    
    pipeline(
      () => on(events, 'screen_interract'),
      reduce((old_canvas, { x, y, screen_id, hand }) => {
        const canvas = copy_canvas(old_canvas)
        
        const ctx = canvas.getContext('2d')
        ctx.font = '30px arial'
        ctx.fillStyle = 'red'
        ctx.beginPath()
        ctx.ellipse(x, y, 10, 10, 0, 0, Math.PI * 2)
        ctx.fill()
        
        update_screen(
          { client, world },
          { screen_id, newCanvas: canvas, oldCanvas: old_canvas }
        )
        
        return canvas
      }, canvas)
    )
```

File: src/index.js (hunk 2 above)
Comment (DeltaEvo) [reply src/index.js:57]: But I think it's unused since you recreate an object

File: src/screen.js [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+        .getContext('2d')
+        .getImageData(
+          128 * (frame % size.width),
+          128 * Math.floor(frame / size.width),
+          128,
+          128
+        )
+      for (let i = 0; i < new_image_data.data.length; i++) {
+        if (new_image_data.data[i] !== old_image_data.data[i]) different = true
+      }
```
Comment (DeltaEvo) [inline src/screen.js:141]: Use buffer.equals

File: src/screen.js [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+  }
+}
+
+export function update_screen(
+  { client, world },
+  { screen_id, newCanvas, oldCanvas }
+) {
+  const { size, start_id } = get_screen(world, screen_id)
+
+  for (let frame = 0; frame < size.width * size.height; frame++) {
```
Comment (DeltaEvo) [inline src/screen.js:119]: Why not create two loops with `x` and `y`

File: src/screen.js [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+    }
+
+    if (different) {
+      const buff = Buffer.alloc(128 * 128, 4)
+      for (let i = 0; i < new_image_data.data.length; i += 4) {
+        const r = new_image_data.data[i]
+        const g = new_image_data.data[i + 1]
+        const b = new_image_data.data[i + 2]
+        buff[i / 4] = nearestMatch(r, g, b)
+      }
```
Comment (DeltaEvo) [inline src/screen.js:151]: Maybe we should use the A8 pixel format to remove that https://github.com/Automattic/node-canvas#image-pixel-formats-experimental ? and avoid to do copies here

File: src/screen.js [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+        icons: [],
+        columns: -128,
+        rows: -128,
+        x: 0,
+        y: 0,
+        data: buff,
+      })
+    }
+  }
+  return { newCanvas }
```
Comment (DeltaEvo) [inline src/screen.js:166]: I think we don't this return :)

File: src/screen.js [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+        entityId: index + start_id,
+        position: frame_pos,
+      })
+    }
+  }
+}
+
+export function update_screen(
+  { client, world },
+  { screen_id, newCanvas, oldCanvas }
```
Comment (DeltaEvo) [inline src/screen.js:115]: `new_canvas`, `old_canvas`

File: src/screen.js [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+  client.write('entity_destroy', {
+    entityIds: Array.from({
+      length: size.width * size.height,
+    }).map((v, index) => start_id + index),
+  })
+}
+
+export function get_screen(world, screen_id) {
+  return world.screens[screen_id]
+}
```
Comment (DeltaEvo) [inline src/screen.js:181]: I think this function is not needed since we can do `world.screens[screen_id]`

File: src/screen.js [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+
+export function copy_canvas(old_canvas) {
+  const new_canvas = createCanvas(old_canvas.width, old_canvas.height)
+  const imageData = old_canvas
```
Comment (DeltaEvo) [inline src/screen.js:185]: `image_data`

File: src/screen.js [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+                screen_id,
+                hand,
+              })
+            }
+          }
+        }
+      }
+    })
+  })
+}
```
Comment (DeltaEvo) [inline src/screen.js:226]: Maybe transform this into an utility function that return the hit or null instead :)

File: src/settings.js [hunk 16]
```diff
@@ -1,5 +1,5 @@
 const { ONLINE_MODE = 'false' } = process.env
 
-export const version = '1.16.3'
+export const version = '1.16.4'
```
Comment (DeltaEvo) [inline src/settings.js:3]: Revert this change :)

**Review by DeltaEvo — CHANGES_REQUESTED** [review]

**Comment by loucass003** [comment]
I Fixed this PR with your feedback

**Comment by loucass003** [comment]
Shoud be all good this time ;)

File: src/player/screen.js [hunk 17]
```diff
@@ -0,0 +1,234 @@
...
+        .getContext('2d')
+        .getImageData(128 * frame_x, 128 * frame_y, 128, 128)
+      if (old_canvas) {
+        const old_image_data = old_canvas
+          .getContext('2d')
+          .getImageData(128 * frame_x, 128 * frame_y, 128, 128)
+        different = !Buffer.from(old_image_data.data.buffer).equals(
+          Buffer.from(new_image_data.data.buffer)
+        )
+      }
```
Comment (DeltaEvo) [inline src/player/screen.js:131]:
```suggestion
      let different = true
      const new_image_data = new_canvas
        .getContext('2d')
        .getImageData(128 * frame_x, 128 * frame_y, 128, 128)
      const old_image_data = old_canvas
          ?.getContext('2d')
          .getImageData(128 * frame_x, 128 * frame_y, 128, 128)
      const different = old_image_data && Buffer.from(old_image_data.data.buffer).equals(
          Buffer.from(new_image_data.data.buffer)
        )
```

**Review by DeltaEvo — APPROVED** [review]

**Pull request merged**